bambi/plots/create_data.py

```diff
@@ -6,124 +6,108 @@
-    user_passed: bool, optional
-        Whether the user passed their own 'conditional' data. Defaults to False.
-
-    Returns
-    -------
-    pd.DataFrame
-        The data for the Conditional Adjusted Comparisons dataframe and or
-        plotting.
-    """
-
-    def _grid_level(condition: ConditionalInfo, contrast: ContrastInfo):
+    def _grid_level():
         """
-        Creates the data for grid-level contrasts by using the covariates passed
-        into the `conditional` arg. Values for the grid are either: (1) computed
-        using a equally spaced grid, mean, and or mode (depending on the covariate
-        dtype), and (2) a user specified value or range of values.
+        Creates a "grid" of data by using the covariates passed into the
+        `conditional` argument. Values for the grid are either: (1) computed
+        using a equally spaced grid, mean, and or mode (depending on the
+        covariate dtype), and (2) a user specified value or range of values.
         """
-        covariates = get_covariates(condition.covariates)
+        covariates = get_covariates(condition_info.covariates)
 
         if user_passed:
-            data_dict = {**condition.conditional}
+            data_dict = {**condition_info.conditional}
         else:
-            main_values = make_main_values(condition.model.data[covariates.main])
+            main_values = make_main_values(condition_info.model.data[covariates.main])
             data_dict = {covariates.main: main_values}
             data_dict = make_group_panel_values(
-                condition.model.data,
+                condition_info.model.data,
                 data_dict,
                 covariates.main,
                 covariates.group,
                 covariates.panel,
-                kind="comparison",
+                kind=kind,
             )
 
-        data_dict[contrast.name] = contrast.values
-        comparison_data = set_default_values(condition.model, data_dict, kind="comparison")
-        # use cartesian product (cross join) to create contrasts
+        data_dict[variable_info.name] = variable_info.values
+        comparison_data = set_default_values(condition_info.model, data_dict, kind=kind)
+        # use cartesian product (cross join) to create pairwise grid
         keys, values = zip(*comparison_data.items())
-        contrast_dict = [dict(zip(keys, v)) for v in itertools.product(*values)]
+        pairwise_grid = pd.DataFrame([dict(zip(keys, v)) for v in itertools.product(*values)])
 
-        return enforce_dtypes(condition.model.data, pd.DataFrame(contrast_dict))
+        # cannot enfore dtypes if 'slopes' because it may remove floating point of dy/dx
+        if kind == "comparisons":
+            pairwise_grid = enforce_dtypes(condition_info.model.data, pairwise_grid)
 
-    def _unit_level(contrast: ContrastInfo):
+        return pairwise_grid
+
+    def _unit_level():
```

Comment:
I would like to avoid functions with no argument that do something depending on a value in an upper-level namespace (i.e. `variable_info` in this case). Could you add the variable_info as an argument to the function?

Reply:
same thing with `_grid_level()` which I've just noticed